File: Marcianitos/src/deepspace/Dice.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.Random;

/**
 *
 * @author mario
 */
class Dice {
    private final float NHANGARSPROB;
    private final float NSHIELDSPROB;
    private final float NWEAPONSPROB;
    private final float FIRSTSHOTPROB;
    
    private Random generator;
    
    Dice (){
        this.NHANGARSPROB = 0.25f;
        this.NSHIELDSPROB = 0.25f;
        this.NWEAPONSPROB = 0.33f;
        this.FIRSTSHOTPROB = 0.5f;
        
        this.generator = new Random();
    }
    
    int initWithNHangars (){
        int retorno = 1;
        
        if (this.generator.nextFloat() < NHANGARSPROB){
            retorno = 0;
        }
        
        return (retorno);
    }
    
    int initWithNWeapons (){
        int retorno = 3;
        
        float num = this.generator.nextFloat();
        
        if (num < NWEAPONSPROB){
            retorno = 1;
            
        } else if (num < 2*NWEAPONSPROB){
            retorno = 2;
        }
        
        return (retorno);
    }    
    
}

File: RUBY/src/Dice.rb
#encoding: UTF-8

module Deepspace
	class Dice
		def initialize
			# Atributos de instancia
			@NHANGARSPROB = 0.25
			@NSHIELDSPROB = 0.25
			@NWEAPONSPROB = 0.33
			@FIRSTSHOTPROB = 0.5
			@generator = Random.new
			
		end
		
		attr_reader :generator
		
		def initWithNHangars
			retorno = 1
			
			if (@generator.rand() < @NHANGARSPROB)
				retorno = 0
			end
			
			return (retorno)
			
		end
		
		def initWithNWeapons
			retorno = 3
			
			num = @generator.rand()
			
			if (num < @NWEAPONSPROB)
				retorno = 1
			elsif (num < 2*@NWEAPONSPROB)
				retorno = 2
			end
			
			return (retorno)
		end
		
		def initWithNShields
			retorno = 1
			
			if (@generator.rand() < @NSHIELDSPROB)
				retorno = 0
			end
			
			return (retorno)
		end
		
		def whoStarts (nPlayers)
			return (@generator.rand(nPlayers))
		end
		
		def firstShot
			if (@generator.rand() < @FIRSTSHOTPROB)
				retorno = GameCharacter::SPACESTATION
			else
				retorno = GameCharacter::ENEMYSTARSHIP
			end
			
			return (retorno)
		end
		
		def spaceStationMoves (speed)
			if (@generator.rand() < speed)
				retorno = true
			else
				retorno = false
			end
			
			return (retorno)
		end
	end
end
